internal/usecase/interactor/asset.go

```diff
@@ -31,95 +30,66 @@ func (i *Asset) Fetch(ctx context.Context, assets []id.AssetID, operator *usecas
 }
 
 func (i *Asset) FindByTeam(ctx context.Context, tid id.TeamID, keyword *string, sort *asset.SortType, p *usecase.Pagination, operator *usecase.Operator) ([]*asset.Asset, *usecase.PageInfo, error) {
-	if err := i.CanReadTeam(tid, operator); err != nil {
-		return nil, nil, err
-	}
-
-	return i.repos.Asset.FindByTeam(ctx, tid, repo.AssetFilter{
-		Sort:       sort,
-		Keyword:    keyword,
-		Pagination: p,
-	})
+	return Run2(
+		ctx, operator, i.repos,
+		Usecase().WithReadableTeams(tid),
+		func() ([]*asset.Asset, *usecase.PageInfo, error) {
+			return i.repos.Asset.FindByTeam(ctx, tid, repo.AssetFilter{
+				Sort:       sort,
+				Keyword:    keyword,
+				Pagination: p,
+			})
+		},
+	)
 }
 
 func (i *Asset) Create(ctx context.Context, inp interfaces.CreateAssetParam, operator *usecase.Operator) (result *asset.Asset, err error) {
-	if err := i.CanWriteTeam(inp.TeamID, operator); err != nil {
-		return nil, err
-	}
-
 	if inp.File == nil {
 		return nil, interfaces.ErrFileNotIncluded
 	}
 
-	tx, err := i.repos.Transaction.Begin()
-	if err != nil {
-		return
-	}
-	defer func() {
-		if err2 := tx.End(ctx); err == nil && err2 != nil {
-			err = err2
-		}
-	}()
-
-	url, err := i.gateways.File.UploadAsset(ctx, inp.File)
-	if err != nil {
-		return nil, err
-	}
-
-	result, err = asset.New().
-		NewID().
-		Team(inp.TeamID).
-		Name(path.Base(inp.File.Path)).
-		Size(inp.File.Size).
-		URL(url.String()).
-		Build()
-	if err != nil {
-		return nil, err
-	}
-
-	if err = i.repos.Asset.Save(ctx, result); err != nil {
-		return
-	}
-
-	tx.Commit()
-	return
+	return Run1(
+		ctx, operator, i.repos,
+		Usecase().
+			WithReadableTeams(inp.TeamID).
+			Transaction(),
+		func() (*asset.Asset, error) {
+			url, err := i.gateways.File.UploadAsset(ctx, inp.File)
+			if err != nil {
+				return nil, err
+			}
+
+			return asset.New().
+				NewID().
+				Team(inp.TeamID).
+				Name(path.Base(inp.File.Path)).
+				Size(inp.File.Size).
+				URL(url.String()).
+				Build()
+		})
```

Comment:
Extracting the logic as a private method would be more readable

Reply:
I don't think it would be that effective because the amount of code would increase...